File: DesignPatterns/Creational/AbstractFactory/Java/CarClient.java
enum CarType {
  MICRO, LUXURY
}

abstract class Car {

  private CarType model;
  private Location location;
  
  public Car(CarType model, Location location) {
    this.model = model;
    this.location = location;
  }

  abstract void construct();

  public CarType getModel(){
    return this.model;
  }

  public void setModel(CarType model){
    this.model = model;
  }

  public Location getLocation(){
    return this.location;
  }

  public void setLocation(Location location) {
    this.location = location;
  }

  public String toString(){
    return "CarModel - " + this.model + " located in - " + this.location;
  }

}

class LuxuryCar extends Car {

  public LuxuryCar(Location location){
    super(CarType.LUXURY, location);
    construct();
  }

  protected void construct(){
    System.out.println("Connecting to LuxuryCar");
  }
}

class MicroCar extends Car {

  public MicroCar(Location location) {
    super(CarType.MICRO, location);
    construct();
  }

  protected void construct(){
    System.out.println("Connecting to MicroCar");
  }
}

enum Location {
  USA, INDIA
}

class IndiaCarFactory {

  public static Car buildCar(CarType model) {
    Car car = null;
    switch(model) {
      case MICRO:
        car = new MicroCar(Location.INDIA);
        break;
      case LUXURY:
        car = new LuxuryCar(Location.INDIA);
        break;
      default:
        break;
    }
    return car;
  }
}

class UsaCarFactory {

  public static Car buildCar(CarType model) {
    Car car = null;
    switch(model) {
      case MICRO:
        car = new MicroCar(Location.USA);
        break;
      case LUXURY:
        car = new LuxuryCar(Location.USA);
        break;
      default:
        break;
    }
    return car;
  }
}

class CarFactory {

  private CarFactory() {

  }

  public static Car buildCar(CarType model){
    Car car = null;

    //We can add location logic here to identify the location dynamically
    Location location = Location.INDIA;

    switch(location) {
      case USA:
        car = UsaCarFactory.buildCar(model);
        break;
      case INDIA:
        car = IndiaCarFactory.buildCar(model);
        break;
      default:
        System.out.println("Not present in your location");
    }
    return car;
  }
}

public class CarClient {
  public static void main(String[] args) {
    System.out.println(CarFactory.buildCar(CarType.MICRO));
    System.out.println(CarFactory.buildCar(CarType.LUXURY));
  }
}

File: DesignPatterns/Creational/AbstractFactory/Ruby/car_client.rb
module CarConstants
  module Type
    MICRO = "MICRO".freeze
    LUXURY = "LUXURY".freeze
  end
  module Location
    USA = "USA".freeze
    INDIA = "INDIA".freeze
  end
end

class Car
  include CarConstants::Location
  include CarConstants::Type

  attr_accessor :model, :location

  def initialize(model, location)
    self.model = model
    self.location = location
  end

  def construct
    raise NotImplementedError
  end

  def to_s
    "CarModel - #{self.model} located in - #{self.location}"
  end
end

class LuxuryCar < Car
  def initialize(location)
    super(LUXURY, location)
    construct()
  end

  def construct
    puts 'Connecting to LuxuryCar'
  end
end

class MicroCar < Car
  def initialize(location)
    super(MICRO, location)
    construct();
  end

  def construct
    puts 'Connecting to MicroCar'
  end
end

class IndiaCarFactory
  def self.build_car(model)
    case model
    when CarConstants::Type::MICRO then MicroCar.new(CarConstants::Location::INDIA)
    when CarConstants::Type::LUXURY then LuxuryCar.new(CarConstants::Location::INDIA)
    end
  end
end

class UsaCarFactory
  def self.build_car(model)
    case model
    when CarConstants::Type::MICRO then MicroCar.new(CarConstants::Location::USA)
    when CarConstants::Type::LUXURY then LuxuryCar.new(CarConstants::Location::USA)
    end
  end
end

class CarFactory
  def self.build_car(model)
    location = CarConstants::Location::INDIA
    case location
    when CarConstants::Location::INDIA then IndiaCarFactory.build_car(model)
    when CarConstants::Location::USA then UsaCarFactory.build_car(model)
    end
  end
end

puts CarFactory.build_car(CarConstants::Type::MICRO)
puts CarFactory.build_car(CarConstants::Type::LUXURY)

